hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -3774,18 +3784,28 @@ private TableGroup createTableGroup(TableGroup parentTableGroup, SqmPath<?> join
 				querySpec.getFromClause().addRoot( tableGroup );
 			}
 			else {
+				final SqlAstJoinType sqlAstJoinType;
+				if ( currentClauseStack.getCurrent() != Clause.SELECT
```

Comment:
Apparently this is not the case, if I remove the `currentClauseStack.getCurrent() != Clause.SELECT` condition I get test failures. In particular: 
* `NotFoundExceptionLogicalOneToOneTest#testQueryAssociationSelection`
* `NotFoundExceptionManyToOneTest#testQueryAssociationSelection`
* `NotFoundIgnoreManyToOneTest#testQueryAssociationSelection`
* `NotFoundIgnoreOneToOneTest#testQueryAssociationSelection`

Reply:
Ok, this is because the upgrade to inner join is only done if a table group already exists for the path. If the table group is created as part of resolving the path, then `prepareForSelection` assumes that the join already uses `inner`, which is kind of wrong now since we changed the default join type.

A HQL query like `select e.notFoundAssociation from MyEntity e` should produce a query like this:

```sql
select *
from my_entity e
join association_table a on e.fk = a.pk
```

I guess with the current state, it will produce a left join now? To fix that, you'd need to call `upgradeToInnerJoinIfNeeded` in the `if ( tableGroup == null ) {` cas in `prepareForSelection`.

Reply:
See my previous comment for how I've handled join "upgrades" to `inner`.